File: week5/python/LongestUnivaluePath_687.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        
        self.max_path = 0
        def getPath(root: TreeNode) -> int:
            if not root:
                return 0
            
            if not root.left and not root.right:
                return 1
            
            cur_path = 0
            max_height = 0
            if root.left:
                height = getPath(root.left)
                if root.val == root.left.val:
                    cur_path += height
                    max_height = height
            
            if root.right:
                height = getPath(root.right)
                if root.val == root.right.val:
                    cur_path += height
                    max_height = max(max_height, height)
            
            self.max_path = max(self.max_path, cur_path)
            return 1 + max_height
        
        getPath(root)
        return self.max_path

File: week5/java/LongestUnivaluePath_687.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    private int maxPath;
    public int longestUnivaluePath(TreeNode root) {
        // int[] maxPath = new int[1];
        findMaxDepth(root);
        return maxPath;
    }
    
    private int findMaxDepth(TreeNode root) {
        if (root == null || (root.left == null && root.right == null)) {
            return 0;
        }
        
        // find max left Univalue depth and record it if same value
        int leftUnivalueDepth = findMaxDepth(root.left) + 1;
        // find max right Univalue depth and record it if same value
        int rightUnivalueDepth = findMaxDepth(root.right) + 1;
        
        int maxCurDepth = 0;
        // int curPath = 0;
        
//         if (root.left != null && root.left.val == root.val) {
//             curPath = leftUnivalueDepth;
//             maxCurDepth = leftUnivalueDepth;
//         } 
        
//         if (root.right != null && root.right.val == root.val) {
//             curPath += rightUnivalueDepth;
//             maxCurDepth = Math.max(maxCurDepth, rightUnivalueDepth);
//         }
        
        if (root.left != null && root.left.val == root.val) {
            if (root.right != null && root.right.val == root.val) {
                maxPath = Math.max(maxPath, leftUnivalueDepth + rightUnivalueDepth);
                maxCurDepth = Math.max(leftUnivalueDepth, rightUnivalueDepth);
            } else {
                maxCurDepth = leftUnivalueDepth;
                maxPath = Math.max(maxPath, leftUnivalueDepth);
            }
            
        } else if (root.right != null && root.right.val == root.val) {
            maxPath = Math.max(maxPath, rightUnivalueDepth);
            maxCurDepth = rightUnivalueDepth;
        }
        
        // update found maxPath
        // maxPath = Math.max(maxPath, curPath);
        
        return maxCurDepth;
    }
}
